src/RSocketClient.h

```diff
@@ -56,13 +60,26 @@ class RSocketClient {
       std::unique_ptr<DuplexConnection> connection,
       folly::EventBase& eventBase,
       SetupParameters setupParameters = SetupParameters(),
-      std::shared_ptr<RSocketResponder> responder = std::shared_ptr<RSocketResponder>(),
-      std::unique_ptr<KeepaliveTimer> keepaliveTimer = std::unique_ptr<KeepaliveTimer>(),
+      std::shared_ptr<RSocketResponder> responder =
+          std::shared_ptr<RSocketResponder>(),
+      std::unique_ptr<KeepaliveTimer> keepaliveTimer =
+          std::unique_ptr<KeepaliveTimer>(),
       std::shared_ptr<RSocketStats> stats = std::shared_ptr<RSocketStats>(),
-      std::shared_ptr<RSocketNetworkStats> networkStats = std::shared_ptr<RSocketNetworkStats>());
+      std::shared_ptr<RSocketNetworkStats> networkStats =
+          std::shared_ptr<RSocketNetworkStats>());
+
+  folly::Future<folly::Unit> resume(
+      std::unique_ptr<ClientResumeStatusCallback> resumeCallback =
+          std::make_unique<DefaultResumeCallback>());
+
+  void disconnect(folly::exception_wrapper ex);
 
  private:
   std::unique_ptr<ConnectionFactory> connectionFactory_;
   std::unique_ptr<RSocketConnectionManager> connectionManager_;
+
+  std::shared_ptr<RSocketStateMachine> stateMachine_;
```

Comment:
Not sure I have the context or understand your question completely. I think the previous implementation was ambiguous on ReactiveSocket ownership.  In this implementation the ConnectionManager takes care of properly closing/destroying the RSocketStateMachines.  It is the responsibility of the client to hold the RSocketClient instance.